### DISCUSSION TITLE
Control of sort order of cape-history

### DISCUSSION THREAD

**Discussion by ashlineldridge** [minad/cape#123] [Q&A] [answered]
I like to use `cape-history` and `consult-history` in different contexts and I also like to use Corfu's [move to minibuffer](https://github.com/minad/corfu?tab=readme-ov-file#transfer-completion-to-the-minibuffer) for moving a `cape-history` completion to the minibuffer. I've noticed that the order of results returned by `cape-history` is different to `consult-history`. I believe this is controlled by Vertico's `vertico-sort-function` variable. For history completion, the nice thing about controlling the sort is so that more recently used commands are shown first by specifying the value `#'vertico-sort-history-length-alpha`. Is control over sort ordering possible with `cape-history`?

**Comment by minad** [comment]
Hi @ashlineldridge! Actually UI sorting should be disabled for both `cape-history` and `consult-history`, so Vertico should not interfere at all and you always get the default history order.

* https://github.com/minad/consult/blob/fe4852280006e61be7f1374d021ee06155ce5a26/consult.el#L4134
* https://github.com/minad/cape/blob/f61da109a9e4491614938c300291060fd8855c1b/cape.el#L391

The sorting is also preserved when moving to the minibuffer. I just tried `consult-history` and `cape-history` in eshell and they behave consistently. So it is not entirely clear to me what you want to achieve, or where you observe an inconsistency. If you want to override sorting forcibly in Vertico you can set the `vertico-sort-override-function` and the equivalent `corfu-sort-override-function`.

**Comment by ashlineldridge** [comment]
Hey! Hmmm, interesting... I get inconsistent results between `cape-history` and `consult-history` - it could be something in my config though. I am able to reproduce this by calling an executable on PATH via it's full path and with no path. E.g.:

```
# Eshell session
$ fd -h
$ /opt/homebrew/bin/fd -h

# Type fd and then M-x cape-history
```

If I follow the above steps, `fd -h` will be shown before `/opt/homebrew/bin/fd -h` even though the latter was called more recently. However, if I move the completion session to the minibuffer, or call `consult-history`, then `/opt/homebrew/bin/fd -h` is shown first.

My preference is for the time order of history to take precedence, which seems to be what's happening with `consult-history` and Vertico (and my settings). I'll look into `corfu-sort-override-function` - I don't currently have either of those overrides set.

**Reply by minad** [reply, marked as answer]
Ah okay, I see. I think the reason is that Corfu explicitly sorts prefix completions earlier. Here you entered the prefix "fd" and "fd -h" takes precedence. See `corfu--move-prefix-candidates-to-front`:

https://github.com/minad/corfu/blob/cdc3e13ad312f5f12b3f78f842fff0b398eb4473/corfu.el#L588

You could override this function with an advice. But generally I find this behavior more aligned with expectations - but maybe not in the case of histories.
